### ISSUE TITLE
Change shadows to subscripts?

### ISSUE THREAD

**Issue by MichaelRFairhurst** [MichaelRFairhurst/wake-compiler#153]
Seems a bit silly I didn't think of this earlier.

But say that we disallow numbers in classnames. Not a huge deal, biggest thing I can think of is

     every ParserVOne is:
     every ParserVTwo is:

being worse than

    every ParserV1 is:
    every ParserV2 is:

But anyway, small cost.

Now you can do

    that(Num1)Equals(Num2) {
        ...
    }

and not only have more readability than Num/$Num, but also (most importantly) have better vertical formatting.

    Num += 1;
    $Num += 1;
    $$Num += 1;

yuck.

    Num1 += 1;
    Num2 += 1;
    Num3 += 1;

ah much better.

Would be cool to support unicode subscripts too, making things very mathematical.

    Numₐ += 4;
    Numₑ += 5;

hmm.

**Comment by RenanStigliani** [comment]
I like the idea of the numbers, but I also think disallowing it in classnames is a huge cost, what about:

```
Num$1 += 1
Num$2 += 1
Num$3 += 1
```

**Comment by IamfromSpace** [comment]
I think I prefer the look of Num$2 to Num2, as it looks more "annotated," rather than just part of the name.

My only concern with Num$3 over $$Num is that you're likely going to cost more key strokes on average.  It's pretty rare to use three of the same class, which is when the two strategies have equal number of strokes involved.

It might be useful to do a combo:
Num or 0$Num
$Num or 1$Num
2$Num
3$Num

Then if you are only using two you have a pretty simple strategy for identifying two, and you have a simple readable count that's the number of '$'

**Comment by MichaelRFairhurst** [comment]
Agreed. Especially as a newcomer to the language, seeing someone suddenly use Num1, Num2, etc, would look using a new part of the standard library rather than using a niche feature. I would expect to be scratching my head at first and then be frustrated when I can't figure out the code and googling for "Num1" turns up nothing.

And I definitely think that well-aligned code is worth extra keystrokes for readability. Consider the following:

```
var $Num = RotationDetector.enqueue(Num).calculateDifferenceBetween(x).run();
var $$Num = RotationDetector.enqueue($Num).calculateDifferenceBetween(y).run();
var $$$Num = RotationDetector.enqueue($$Num).calculateDifferenceBetween(y).run();
```

Spot the "Bug" in this code? No, its not terribly hard to find. But I told you it had one. What about

```
var Num$2 = RotationDetector.enqueue(Num$1).calculateDifferenceBetween(x).run();
var Num$3 = RotationDetector.enqueue(Num$2).calculateDifferenceBetween(y).run();
var Num$4 = RotationDetector.enqueue(Num$3).calculateDifferenceBetween(y).run();
```

Its pretty easy now to spot the differing columns, and scan through them. You quickly see 234 on the left, 123 in the middle, and then x y y stands out like a sore thumb. This is not as clear when the columns aren't perfectly aligned.

A also don't like the idea of people doing

```
var $Num   = RotationDetector.enqueue(Num  ).calculateDifferenceBetween(x).run();
var $$Num  = RotationDetector.enqueue($Num ).calculateDifferenceBetween(y).run();
var $$$Num = RotationDetector.enqueue($$Num).calculateDifferenceBetween(y).run();
```

If we can find a way to make this code readable without bringing whitespace preference into it that would definitely be ideal.

I'm not a huge fan of the prefix style "0$Num" because I think it puts the least important part first. But maybe I'm being weird?

In terms of quantifying the usefulness of this feature, something that stands out to me is that the most logical next request for this feature is to allow arbitrary identifiers, ie,

Num$a
Num$b

but now that Wake has type inference, you wonder why you didn't simply say "var a = 5". Which does make me wonder why I ever even added support Num and $Num. The only reason I can think of is that you're defining method argument so you don't have the luxury of type inference. Still though, why wouldn't you use 'Num a' instead of 'Num$a'?

Going back to the basics, shadows were mostly just made to solve the moment of when you want to write this code but feel guilty about it:

```
setHeight(int a) {
    height = a;
}
```

So in Wake you canuse $Num instead of 'Num a', guilt free.

But the minute we have two variables,

   setDimensions($Num, $$Num)

is already probably not a great idea and I'm not sure its worth the hassle to support

   setDimensions(Num$1, Num$2)

when people really should at this point be using

   setDimensions(Num w, Num h)

So it strikes me as useful in certain algorithmic conditions where you have lots of throwaway numbers, for instance I implemented MD5 in wake recently. At one point I opted for shadows which felt nice:

```
   //Initialize variables:
    var Num = 0x67452301;
    var $Num = 0xefcdab89;
    var $$Num = 0x98badcfe;
    var $$$Num = 0x10325476;
```

and later on I used lowercase vars which felt nicer mostly due to the alignment and the fact that I wasn't gonna type out $$$$$$Num haha

```
        //Initialize hash value for this chunk:
        var a = Num;
        var b = $Num;
        var c = $$Num;
        var d = $$$Num;
```

Might have been nice to be able to use 'subscripts' like

```
        var Num$5 = Num$1;
        var Num$6 = Num$2;
        var Num$7 = Num$3;
        var Num$8 = Num$4;
```

So I think its an improvement but a mild one under average conditions.

 Maybe shadows in general should be scrapped in the goal of keeping the language small & simple...

**Comment by IamfromSpace** [comment]
Well, I think honestly the use of more than three shadows of the same type is a clear code smell.

For example, a good case for shadows might be:

```
sum(Num, $Num) {
  return Num + $Num;
}
```

I think this is super valid and is really readable.  Tell the reader--look they're are just numbers, nothing else about them matters.

Why not a triple sum function with three shadows?  Because we're much better off working with an array.

```
sum(Num[]) {
  var Num running = 0;
  for (i=0;i <Num[].length;i++) {
    running += Num[i];
  }
  return running;
}
```

Yay, we've solved our generic case.

Because it's a (likely) code smell, we shouldn't worry too much about the third or fourth shadow--but who knows, maybe someone will do something cool with it given the tools.

But I still think shadows are good, especially when multiple types are involved and the only important identifier is the type.

```
makeBullet(Num, Text) {
  return Num.toString() + ". " + Text;
}
```

In fairness in this example, if you wanted to make the delimeter a param, it would probably be better to name the two Texts than to shadow them.

My point is, I still really like shadows, but they are likely a code smell when over used; it indicates and in-abstracted pattern.

Given all that, I'm in favor of "Num" and "$Num."

If we want to make deeper shadows simple to use for the rare case it isn't a code smell, array-ishness is likely the best bet.
